### ISSUE TITLE
Cannot use process substitution for QuickFix mode

### ISSUE THREAD

**Issue by strokirk** [neovim/neovim#21756]
### Describe the bug

Attempting to open neovim in QuickFix mode using process substitution does not work, i.e. `nvim -q <(rg --vimgrep foo)`.

The program stops with the error `E40: Can't open errorfile /dev/fd/12` printed in the logs.

Using `git bisect` it seems have been introduced by https://github.com/neovim/neovim/pull/18375, but I'm not yet sure which line of code is the issue.

### Steps to reproduce

```
git checkout cce73621;
make;
VIMRUNTIME=runtime ./build/bin/nvim --clean -q <(rg ran-deps --vimgrep) -Vlog;
tail log;

# Outputs: E40: Can't open errorfile /dev/fd/12
```

### Expected behavior

Vim opens with the QuickFix list populated.

### Neovim version (nvim -v)

NVIM v0.9.0-dev-589+gcce736218

### Vim (not Nvim) behaves the same?

No

### Operating system/version

macOS 13.0.1

### Terminal name/version

iTerm2 3.4.19

### $TERM environment variable

xterm-256color

### Installation

build from repo

**Comment by zeertzjq** [comment]
Cannot reproduce on Linux with bash or zsh. What shell are you using?

**Comment by strokirk** [comment]
@zeertzjq This happens for me both in `bash` and `zsh` (`GNU bash, version 3.2.57(1)-release (arm64-apple-darwin22)` and `zsh 5.9 (arm-apple-darwin22.1.0)`).

**Comment by strokirk** [comment]
The issue seems to happen in the [`os_fopen` call in `qf_setup_state`](https://github.com/bfredl/neovim/blob/9fdcbbb4063daa125e420e0ffe9dae6801c264bc/src/nvim/quickfix.c#L1023), which fails.

**Comment by strokirk** [comment]
Interestingly, the file opening error only occurs after the line `ui_client_run` has been run.

Adding these lines before and after `ui_client_run` shows how the file seems to be able to be opened before, but not after.

```c
  if (params.edit_type == EDIT_QF && params.use_ef != NULL) {
    set_string_option_direct("ef", -1, params.use_ef, OPT_FREE, SID_CARG);
    if (os_fopen((char *)p_ef, "r") == NULL) {
      semsg("DBG:OpenError:%s", (char *)p_ef);
    } else {
      semsg("DBG:Open:%s", (char *)p_ef);
    }
 }
 ```

**Comment by zeertzjq** [comment]
> Adding these lines before and after `ui_client_run`

`ui_client_run` does not return, so lines after `ui_client_run` don't do anything.

**Comment by zeertzjq** [comment]
Wait did you test on latest commit?

**Comment by strokirk** [comment]
@zeertzjq Yes, still an issue on 69c71c4ab44d0bf9fe5f6b6066ad5c269d1507b8. 

Unfortunately I'm not very experienced at C debugging - trying to use `lldb` or a simple `printf` doesn't really work, but I've found `semsg` to produce some output at least. (Likely the wrong approach!)

```
❯ VIMRUNTIME=runtime ./build/bin/nvim --clean -q <(rg ran-deps --vimgrep) -V/tmp/log
DBG:Open:/dev/fd/12%

~/dev/neovim master*                                                                             22:28:56
❯ tail /tmp/log
chdir(/Users/dan/dev/neovim/runtime/plugin/)
chdir(/Users/dan/dev/neovim)
finished sourcing runtime/plugin/nvim.lua
not found in 'packpath': "pack/*/start/*"
not found in 'packpath': "start/*"
not found in runtime path: "plugin/**/*.vim"
not found in runtime path: "plugin/**/*.lua"
                                            E40: Can't open errorfile /dev/fd/12

DBG:Open:/dev/fd/12%
```

**Comment by gpanders** [comment]
Works fine on macOS with fish:

```
nvim -q (rg --vimgrep foo | psub)
```

**Comment by erw7** [comment]
Looking at the following part of `libuv`, it appears that on MacOS, the child process (server) does not inherit the `fd` of the parent process (tui).

https://github.com/libuv/libuv/blob/ee206367d4e8ebb70455665de78b8309220fd7d0/src/unix/process.c#L462-L475
```c
  /* Set flags for spawn behavior
   * 1) POSIX_SPAWN_CLOEXEC_DEFAULT: (Apple Extension) All descriptors in the
   *    parent will be treated as if they had been created with O_CLOEXEC. The
   *    only fds that will be passed on to the child are those manipulated by
   *    the file actions
   * 2) POSIX_SPAWN_SETSIGDEF: Signals mentioned in spawn-sigdefault in the
   *    spawn attributes will be reset to behave as their default
   * 3) POSIX_SPAWN_SETSIGMASK: Signal mask will be set to the value of
   *    spawn-sigmask in attributes
   * 4) POSIX_SPAWN_SETSID: Make the process a new session leader if a detached
   *    session was requested. */
  flags = POSIX_SPAWN_CLOEXEC_DEFAULT |
          POSIX_SPAWN_SETSIGDEF |
          POSIX_SPAWN_SETSIGMASK;
```

Therefore, I think it is necessary to check all file names with the command line option (such as `-u`) that takes the file name, and if it is `/dev/fd/n`, then the same process as the following `stdin` process should be performed so that the child process inherits the `fd`. It is quite troublesome to change `n` in `/dev/fd/n` to `fd` which is passed to the child process side, etc. It may be easier to change `libuv` so that this `flags` can be specified.

https://github.com/neovim/neovim/blob/aec4b476c5689f032103ab3e0ca068fc2af8bbba/src/nvim/event/libuv_process.c#L52-L57

**Comment by erw7** [comment]
The `libuv` on linux does not use `posix_spawn`. Then `fish` creates a temporary file. Therefore, this problem does not occur with them.

**Comment by zdog234** [comment]
Sorry, most of this stuff is going over my head. Is there a workaround for zsh on OSX?

**Comment by zdog234** [comment]
Snippet that I put at `~/.local/bin/nvimq`:

(note: this has performance issues, because it writes all of the data into a file on disk before starting `nvim`)

```python
#!/usr/bin/env python3
import argparse
import io
import subprocess as sp
import sys
import tempfile

parser = argparse.ArgumentParser()
parser.add_argument(
    "file",
    nargs="?",
    type=argparse.FileType("r"),
    default=sys.stdin,
)
file: io.TextIOBase = parser.parse_args().file

with tempfile.NamedTemporaryFile(mode="w+", encoding="utf-8") as tmp:
    tmp.writelines(file)
    tmp.flush()
    sys.exit(sp.call(["nvim", "-q", tmp.name, "+cw"]))
```

usage:

```shell
nvimq <(rg --vimgrep foo)
rg --vimgrep bar | nvimq
```

**Comment by carlocab** [comment]
> Sorry, most of this stuff is going over my head. Is there a workaround for zsh on OSX?

A workaround for `zsh` is to do
```zsh
nvim -q =(rg --vimgrep foo)
```
From `man zshexpn`:
```
If =(...) is used instead of <(...), then the file passed as an argument will be the name of a temporary file containing the
output of the list process.  This may be used instead of the < form for a program that expects to lseek (see lseek(2)) on the
input file.
```
Basically it does what your `nvimq` script does.

**Comment by strokirk** [comment]
Can verify that the workaround @carlocab suggests in `zsh` works great!

**Comment by justinmk** [comment]
> the child process (server) does not inherit the `fd` of the parent process (tui).

Inverse of https://github.com/libuv/libuv/pull/3856 / https://github.com/libuv/libuv/pull/5100 ?
